design/jwt-verification-design.md

```diff
@@ -0,0 +1,205 @@
+Contour will also add a CDS cluster for the remote JWKS, as required by [the Envoy configuration](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto#envoy-v3-api-msg-extensions-filters-http-jwt-authn-v3-remotejwks).
+
+
+## Detailed Design
+The detailed structure of the new fields is shown via YAML below:
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: jwt-verification
+spec:
+  virtualhost:
+    fqdn: example.com
+    tls:
+      secretName: tls-cert
+    jwtProviders:
+      - 
+        # name is a unique name for the provider.
+        name: provider-1
+        # issuer (optional) must match the "iss" field in the JWT.
+        # If not specified, the "iss" field is not checked.
+        issuer: foo.com
+        # audiences (optional) allowlist for the "aud" field in the JWT.
+        # If not specified, the "aud" field is not checked.
+        audiences:
+          - audience-1
+          - audience-2
+        # remoteJWKS is an HTTP endpoint that returns the JWKS
+        # to use to verify the JWT signature.
+        remoteJWKS:
+          httpURI:
+            uri: https://example.com/jwks.json
+            timeout: 1s
+          # cacheDuration is how long to cache the fetched JWKS
+          # locally.
+          cacheDuration: 5m
+        # localJWKS can be used instead of remoteJWKS and defines
+        # an in-cluster secret containing the JWKS for this provider.
+        localJWKS:
+          secretName: my-jwks
+          key: jwks.json
+  routes:
+    # This route specifies jwtProvider: provider-1, which requires requests
+    # to have a JWT has issuer=foo.com, audience of either "audience-1" or 
+    # "audience-2", and signature must be able to be verified using the JWKS
+    # at https://example.com/jwks.json).
+    - conditions:
+        - prefix: /
+      jwtProvider: provider-1
+      services:
+        - name: s1
+          port: 80
+    # This route does *not* have a jwtProvider defined, so requests
+    # with paths starting with /js are excluded from JWT verification.
+    # Note that Contour orders routes such that longer prefixes take
+    # priority over shorter prefixes, so requests starting with /js will
+    # be handled by this route rather than the previous catch-all route.
+    - conditions:
+        - prefix: /js
+      services:
+        - name: s1
+          port: 80
+```
+
+It is worth highlighting some of the configuration options that Envoy's filter has, that Contour will *not* expose, at least not initially:
+- non-default extract locations for the JWT (the default locations are (1) the `Authorization` header using the Bearer schema; and (2) the `access_token` query parameter, in that order).
+- complex rule requirements (e.g. `requires_any`, `requires_all`)
```

Comment:
Just a question to clarify to myself: we will use simple requirement rules such as below?

```yaml
- match:
    prefix: /
  requires: { provider_name: provider-A }
```

Reply:
Yep, that's the plan.